backend/pcloud/chunkwriter.go

```diff
@@ -0,0 +1,337 @@
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"sync"
+
+	"github.com/rclone/rclone/backend/pcloud/api"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/lib/rest"
+	"golang.org/x/sync/errgroup"
+)
+
+type chunkWriter struct {
+	client    *rest.Client
+	fs        *Fs
+	chunkSize int64
+	src       fs.ObjectInfo
+	remote    string
+	fd        int64
+	fileID    int64
+
+	mu        sync.RWMutex
+	byteCount int64
+}
+
+func (c *chunkWriter) GetBytesCount() int64 {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.byteCount
+}
+
+func (c *chunkWriter) AddBytesCount(count int64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.byteCount = c.byteCount + count
+}
+
+func newChunkWriter(
+	ctx context.Context,
+	remote string,
+	src fs.ObjectInfo,
+	srcFs *Fs,
+) (info fs.ChunkWriterInfo, writer *chunkWriter, err error) {
+	client, err := srcFs.newSingleConnClient(ctx)
+	if err != nil {
+		return info, writer, fmt.Errorf("create client: %w", err)
+	}
+	// init an empty file
+	leaf, directoryID, err := srcFs.dirCache.FindPath(ctx, remote, true)
+	if err != nil {
+		return info, writer, fmt.Errorf("resolve src: %w", err)
+	}
+	openResult, err := fileOpenNew(ctx, client, srcFs, directoryID, leaf)
+	if err != nil {
+		return info, writer, fmt.Errorf("open file: %w", err)
+	}
+
+	info = fs.ChunkWriterInfo{
+		ChunkSize:         int64(srcFs.opt.ChunkSize),
+		Concurrency:       srcFs.opt.UploadConcurrency,
+		LeavePartsOnError: true,
+	}
+	writer = &chunkWriter{
+		client:    client,
+		fs:        srcFs,
+		chunkSize: info.ChunkSize,
+		src:       src,
+		remote:    remote,
+		fd:        openResult.FileDescriptor,
+		fileID:    openResult.Fileid,
+	}
+
+	return info, writer, nil
+}
+
+// Abort implements fs.ChunkWriter.
+func (c *chunkWriter) Abort(ctx context.Context) error {
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open aborted: %w", err)
+	}
+	if err := o.Remove(ctx); err != nil {
+		return fmt.Errorf("delete aborted: %w", err)
+	}
+	return nil
+}
+
+// Close implements fs.ChunkWriter.
+func (c *chunkWriter) Close(ctx context.Context) error {
+	// In case the file existed previosly and was larger than the newly written file, we need to truncate it to the right size.
+	byteCount := c.GetBytesCount()
+	fs.Debugf(c.src, "truncate target to %d bytes", byteCount)
+	if _, err := c.fileTruncate(ctx, byteCount); err != nil {
+		return fmt.Errorf("truncate file: %w", err)
+	}
+
+	// close fd
+	if _, err := c.fileClose(ctx); err != nil {
+		return fmt.Errorf("close fd: %w", err)
+	}
+
+	// There seems to be no option in the fileops API to set the modtime.
+	// Therefore, we apply the modtime via the regular file API.
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open newly copied object: %w", err)
+	}
+	if err := o.SetModTime(ctx, c.src.ModTime(ctx)); err != nil {
+		return fmt.Errorf("set modtime: %w", err)
+	}
+
+	return nil
+}
+
+// WriteChunk implements fs.ChunkWriter.
```

Comment:
This function looks redundant now? Perhaps just rename `writeChunk`

Reply:
Not exactly, as you can see from the reader interface. 
The methods are both in use:
- `WriteChunk` is the exported method fullfilling the `ChunkWriter` interface using `io.ReadSeeker`
- `writeChunk` is called by the `fs.Update` while only having a `io.Reader` at hand

Reply:
I added a comment.